plugins/chain_plugin/trx_finality_status_processing.cpp

```diff
@@ -40,82 +47,130 @@ namespace eosio::chain_apis {
-            _storage.modify( iter, [&block_id,&block_timestamp]( finality_status_object& obj ) {
-               obj.block_id = block_id;
-               obj.block_timestamp = block_timestamp;
-               obj.forked_out = false;
-            } );
-         }
-         else {
-            _storage.insert(
-               finality_status_object{.trx_id = trx_id,
-                                      .trx_expiry = std::get<1>(trx_tuple)->expiration(),
-                                      .received = now,
-                                      .block_id = block_id,
-                                      .block_timestamp = block_timestamp});
-         }
+      if (!trace->producer_block_id) {
+         _speculative_trxs.push_back(trace->id);
       }
-      if (modified) {
-         const auto& indx = _storage.index().get<by_status_expiry>();
 
-         // find the lowest value successful block
-         auto success_iter = indx.lower_bound(boost::make_tuple(true, fc::time_point{}));
-         if (success_iter != indx.cend()) {
-            _earliest_tracked_block_id = success_iter->block_id;
-         }
+      if(ensure_storage()) {
+         modified = true;
+      }
+
+      const auto& trx_id = trace->id;
+      auto iter = _storage.find(trx_id);
+      if (iter != _storage.index().cend()) {
+         _storage.modify( iter, [&block_id,&block_timestamp]( finality_status_object& obj ) {
+            obj.block_id = block_id;
+            obj.block_timestamp = block_timestamp;
+            obj.forked_out = false;
+         } );
+      }
+      else {
+         _storage.insert(
+            finality_status_object{.trx_id = trx_id,
+                                   .trx_expiry = ptrx->expiration(),
+                                   .received = now,
+                                   .block_id = block_id,
+                                   .block_timestamp = block_timestamp});
       }
 
-      if (bsp) {
-         _last_proc_block_num = chain::block_header::num_from_id(_head_block_id);
+      if (modified || _earliest_tracked_block_id == chain::block_id_type{}) {
+         determine_earliest_tracked_block_id();
       }
    }
 
+   void trx_finality_status_processing_impl::signal_accepted_block( const chain::block_state_ptr& bsp ) {
+      // if this block had any transactions, then we have processed everything we need to already
+      if (bsp->id == _head_block_id) {
+         return;
+      }
+
+      _head_block_id = bsp->id;
+      _head_block_timestamp = bsp->block->timestamp;
+
+      const auto head_block_num = chain::block_header::num_from_id(_head_block_id);
+      if (head_block_num <= _last_proc_block_num) {
+         handle_rollback();
+      }
+
+      const fc::time_point now = fc::time_point::now();
+      bool status_expiry = status_expiry_of_trxs(now);
+      if (status_expiry) {
+         determine_earliest_tracked_block_id();
+      }
+
+      // if this approve block was proceeded by speculative transactions that had
```

Comment:
Update comment. `// if this approve block was proceeded by speculative transactions then we produced the block, update trx state.`

Reply:
And spelling.  I'm fairly sure you meant 'preceded'.

Reply:
Addressed: 4399330